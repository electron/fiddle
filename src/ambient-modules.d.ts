declare module 'namor';
declare module 'shell-path';
