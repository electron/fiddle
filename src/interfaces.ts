import { InstallState } from '@vertedinde/fiddle-core';

export type Files = Map<string, string>;

export type FileTransform = (files: Files) => Promise<Files>;

export enum VersionSource {
  remote = 'remote',
  local = 'local',
}

export enum GistActionType {
  publish = 'Publish',
  update = 'Update',
  delete = 'Delete',
}

export enum GistActionState {
  publishing = 'publishing',
  updating = 'updating',
  deleting = 'deleting',
  none = 'none',
}

export interface Version {
  version: string;
  name?: string;
  localPath?: string;
  node?: string;
}

export enum RunResult {
  SUCCESS = 'success', // exit code === 0
  FAILURE = 'failure', // ran, but exit code !== 0
  INVALID = 'invalid', // could not run
}

export interface RunnableVersion extends Version {
  state: InstallState;
  source: VersionSource;
  downloadProgress?: number;
}

export const enum ElectronReleaseChannel {
  stable = 'Stable',
  beta = 'Beta',
  nightly = 'Nightly',
}

export interface SetFiddleOptions {
  filePath?: string;
  templateName?: string;
  gistId?: string;
}

export interface SetUpMenuOptions {
  acceleratorsToBlock?: BlockableAccelerator[] | null;
  activeTemplate?: string | null;
}

export interface SetupRequest {
  fiddle?: SetFiddleOptions;
  version?: string;
  showChannels: ElectronReleaseChannel[];
  hideChannels: ElectronReleaseChannel[];
  useObsolete?: boolean;
}

export interface BisectRequest {
  setup: SetupRequest;
  goodVersion: string;
  badVersion: string;
}

export interface TestRequest {
  setup: SetupRequest;
}

export interface OutputEntry {
  text: string;
  timeString: string;
  isNotPre?: boolean;
}

export interface OutputOptions {
  bypassBuffer?: boolean;
  isNotPre?: boolean;
}

export interface GenericDialogOptions {
  type: GenericDialogType;
  ok: string;
  cancel?: string;
  wantsInput: boolean;
  defaultInput?: string;
  label: string | JSX.Element;
  placeholder?: string;
}

export interface Contributor {
  url: string;
  api: string;
  login: string;
  avatar: string;
  name: string;
  bio: string;
  location: string;
}

export interface Templates {
  [index: string]: string | Templates;
}

export const enum GenericDialogType {
  'confirm' = 'confirm',
  'warning' = 'warning',
  'success' = 'success',
}

export type EditorId = `${string}.${'js' | 'html' | 'css'}`;

export type EditorValues = Record<EditorId, string>;

// main.js gets special treatment: it is required as the entry point
// when we run fiddles or create a package.json to package fiddles.
export const MAIN_JS = 'main.js';

export const PACKAGE_NAME = 'package.json';

export type ArrowPosition = 'top' | 'left' | 'bottom' | 'right';

export const enum BlockableAccelerator {
  save = 'save',
  saveAs = 'saveAs',
}
