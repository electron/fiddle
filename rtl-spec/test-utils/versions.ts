import {
  ElectronReleaseChannel,
  InstallState,
  VersionSource,
} from '../../src/interfaces';
import { VersionsMock } from '../../tests/mocks/electron-versions';
import { StateMock } from '../../tests/mocks/state';

const { missing } = InstallState;
const { remote } = VersionSource;

export const mockVersion1 = {
  source: remote,
  state: missing,
  version: '26.0.0',
};

export const mockVersion2 = {
  source: remote,
  state: missing,
  version: '28.0.0-unsupported',
};

/**
 * Initializes the app state with our mock versions.
 */
export function prepareAppState() {
  const { state: appState } = window.app;

  const { mockVersions } = new VersionsMock();

  (appState as unknown as StateMock).initVersions('2.0.2', {
    ...mockVersions,
    [mockVersion1.version]: { ...mockVersion1 },
    [mockVersion2.version]: { ...mockVersion2 },
  });

  appState.channelsToShow = [
    ElectronReleaseChannel.stable,
    ElectronReleaseChannel.beta,
  ];

  return appState;
}
