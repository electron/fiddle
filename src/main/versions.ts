import * as path from 'node:path';

import { ElectronVersions, Installer, SemVer } from '@electron/fiddle-core';
import { IpcMainEvent, app } from 'electron';
import * as fs from 'fs-extra';

import releases from '../../static/releases.json';
import { InstallState, Version } from '../interfaces';
import { IpcEvents } from '../ipc-events';
import { ipcMainManager } from './ipc';

let knownVersions: ElectronVersions;

/**
 * Helper to check if this version is from a released major branch.
 *
 * This way when we have a local version of Electron like '999.0.0'
 * we'll know to not try & download 999-x-y.zip from GitHub :D
 *
 * @param {number} major - Electron major version number
 * @returns {boolean} true if there are releases with that major version
 */
export function isReleasedMajor(major: number): boolean {
  return knownVersions.inMajor(major).length > 0;
}

export function getOldestSupportedMajor(): number | undefined {
  const NUM_BRANCHES = parseInt(process.env.NUM_STABLE_BRANCHES || '');

  if (!Number.isNaN(NUM_BRANCHES)) {
    return knownVersions.stableMajors.slice(-NUM_BRANCHES)[0];
  }

  return knownVersions.supportedMajors[0];
}

export function getLatestStable(): SemVer | undefined {
  return knownVersions.latestStable;
}

export function getReleasedVersions(): Array<Version> {
  // Don't support anything older than 0.30 (Aug 2015).
  // The oldest version known to releases.json.org is 0.20,
  // Pre-0.24.0 versions were technically 'atom-shell' and cannot
  // be downloaded with @electron/get.
  // TODO(dsanders11): upstream this logic to @electron/fiddle-core
  return knownVersions.versions
    .filter((ver) => !ver.version.startsWith('0.2'))
    .map(({ version }) => ({ version }));
}

/**
 * Gets the current state of a specific version
 * Valid local electron builds are marked as `installed`
 *
 * @param {Version} ver
 * @returns {InstallState}
 */
export function getLocalVersionState(ver: Version): InstallState {
  const { localPath } = ver;
  if (localPath !== undefined) {
    const dir = Installer.getExecPath(localPath);
    if (fs.existsSync(dir)) {
      return InstallState.installed;
    }
  }

  return InstallState.missing;
}

export async function fetchVersions(): Promise<Version[]> {
  await knownVersions.fetch();
  return getReleasedVersions();
}

export async function setupVersions() {
  knownVersions = await ElectronVersions.create(
    {
      versionsCache: path.join(app.getPath('userData'), 'releases.json'),
    },
    {
      initialVersions: releases,
    },
  );

  ipcMainManager.handle(
    IpcEvents.IS_RELEASED_MAJOR,
    (_: IpcMainEvent, version: number) => isReleasedMajor(version),
  );
  ipcMainManager.handle(IpcEvents.FETCH_VERSIONS, (_: IpcMainEvent) =>
    fetchVersions(),
  );
  ipcMainManager.on(IpcEvents.GET_LATEST_STABLE, (event) => {
    event.returnValue = getLatestStable();
  });
  ipcMainManager.on(IpcEvents.GET_LOCAL_VERSION_STATE, (event, ver) => {
    event.returnValue = getLocalVersionState(ver);
  });
  ipcMainManager.on(IpcEvents.GET_OLDEST_SUPPORTED_MAJOR, (event) => {
    event.returnValue = getOldestSupportedMajor();
  });
  ipcMainManager.on(IpcEvents.GET_RELEASED_VERSIONS, (event) => {
    event.returnValue = getReleasedVersions();
  });
  ipcMainManager.handle(
    IpcEvents.GET_RELEASE_INFO,
    (_: IpcMainEvent, version) => knownVersions.getReleaseInfo(version),
  );

  return knownVersions;
}
