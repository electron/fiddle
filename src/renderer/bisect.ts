import { RunnableVersion } from '../interfaces';

export class Bisector {
  public revList: Array<RunnableVersion>;
  public minRev: number;
  public maxRev: number;
  private pivot: number;

  constructor(revList: Array<RunnableVersion>) {
    this.getCurrentVersion = this.getCurrentVersion.bind(this);
    this.continue = this.continue.bind(this);
    this.calculatePivot = this.calculatePivot.bind(this);

    this.revList = revList;
    this.minRev = 0;
    this.maxRev = revList.length - 1;
    this.calculatePivot();
  }

  public getCurrentVersion() {
    return this.revList[this.pivot];
  }

  public continue(isGoodVersion: boolean) {
    let isBisectOver = false;
    if (this.maxRev - this.minRev <= 1) {
      isBisectOver = true;
    }

    if (isGoodVersion) {
      const upPivot = Math.floor((this.maxRev - this.pivot) / 2) + this.pivot;
      this.minRev = this.pivot;
      if (upPivot !== this.maxRev && upPivot !== this.pivot) {
        this.pivot = upPivot;
      } else {
        isBisectOver = true;
      }
    } else {
      const downPivot =
        Math.floor((this.pivot - this.minRev) / 2) + this.minRev;
      this.maxRev = this.pivot;
      if (downPivot !== this.minRev && downPivot !== this.pivot) {
        this.pivot = downPivot;
      } else {
        isBisectOver = true;
      }
    }

    if (isBisectOver) {
      return [this.revList[this.minRev], this.revList[this.maxRev]];
    } else {
      return this.revList[this.pivot];
    }
  }

  private calculatePivot() {
    this.pivot = Math.floor((this.maxRev - this.minRev) / 2);
  }
}
