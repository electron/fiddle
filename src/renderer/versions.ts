import { InstallState, Installer } from '@vertedinde/fiddle-core';
import * as fs from 'fs-extra';
import semver from 'semver';

import releasesJSON from '../../static/releases.json';
import {
  ElectronReleaseChannel,
  RunnableVersion,
  Version,
  VersionSource,
} from '../interfaces';
import { normalizeVersion } from '../utils/normalize-version';

/**
 * Returns a sensible default version string.
 *
 * @param {Array<RunnableVersion>} knownVersions
 * @returns {string}
 */
export function getDefaultVersion(versions: RunnableVersion[]): string {
  const key = localStorage.getItem('version');
  if (key && versions.some(({ version }) => version === key)) return key;

  // newest stable release
  const latestStable = versions
    .filter((ver) => !ver.version.includes('-')) // stable
    .map((ver) => semver.parse(ver.version))
    .sort((a, b) => -semver.compare(a!, b!))
    .shift();
  if (latestStable) return latestStable.version;

  // how do we not have a stable version listed?
  throw new Error('Corrupted version data');
}

/**
 * Return the release channel for a given input
 * version.
 *
 * @param {Version | string} input
 * @returns {ElectronReleaseChannel}
 */
export function getReleaseChannel(
  input: Version | string,
): ElectronReleaseChannel {
  const tag = typeof input === 'string' ? input : input.version || '';

  if (tag.includes('beta') || tag.includes('alpha')) {
    return ElectronReleaseChannel.beta;
  }

  if (tag.includes('nightly')) {
    return ElectronReleaseChannel.nightly;
  }

  // Must be a stable version, right?
  return ElectronReleaseChannel.stable;
}

export const enum VersionKeys {
  local = 'local-electron-versions',
  known = 'known-electron-versions',
}

/**
 * Retrieve Electron versions from localStorage.
 *
 * @param {VersionKeys} key
 * @param {() => Array<Version>} fallbackMethod
 * @returns {Array<Version>}
 */
function getVersions(
  key: VersionKeys,
  fallbackMethod: () => Array<Version>,
): Array<Version> {
  const fromLs = window.localStorage.getItem(key);

  if (fromLs) {
    try {
      let result: Array<Version> = JSON.parse(fromLs);

      if (!isExpectedFormat(result)) {
        // Known versions can just be downloaded again.
        if (key === VersionKeys.known) {
          throw new Error(
            `Electron versions in LS does not match expected format`,
          );
        }

        // Local versions are a bit more tricky and might be in an old format (pre 0.5)
        result = migrateVersions(result);
        saveLocalVersions(result);
      }

      return result;
    } catch (error) {
      console.warn(
        `Parsing local Electron versions failed, returning fallback method.`,
      );
    }
  }

  return fallbackMethod();
}

/**
 * Save an array of GitHubVersions to localStorage.
 *
 * @param {VersionKeys} key
 * @param {Array<Version} versions
 */
function saveVersions(key: VersionKeys, versions: Array<Version>) {
  const stringified = JSON.stringify(versions);
  window.localStorage.setItem(key, stringified);
}

/**
 * Gets the current state of a specific version
 * Valid local electron builds are marked as `installed`
 *
 * @param {Version} ver
 * @returns {InstallState}
 */
export function getVersionState(ver: Version): InstallState {
  const { localPath } = ver;
  if (localPath !== undefined) {
    const dir = Installer.getExecPath(localPath);
    if (fs.existsSync(dir)) {
      return InstallState.installed;
    }
  }

  return InstallState.missing;
}

export function makeRunnable(ver: Version): RunnableVersion {
  const ret: RunnableVersion = {
    ...ver,
    version: normalizeVersion(ver.version),
    source: Boolean(ver.localPath) ? VersionSource.local : VersionSource.remote,
    state: getVersionState(ver),
  };

  return ret;
}

/**
 * Return both known as well as local versions.
 *
 * @returns {Array<RunnableVersion>}
 */
export function getElectronVersions(): Array<RunnableVersion> {
  const versions = [...getReleasedVersions(), ...getLocalVersions()];
  return versions.map((ver) => makeRunnable(ver));
}

/**
 * Add a version to the local versions
 *
 * @param {Version} input
 * @returns {Array<Version>}
 */
export function addLocalVersion(input: Version): Array<Version> {
  const versions = getLocalVersions();

  if (!versions.find((v) => v.localPath === input.localPath)) {
    versions.push(input);
  }

  saveLocalVersions(versions);

  return versions;
}

/**
 * Get the Version (if any) that is located at localPath.
 *
 * @param {string} folderPath
 * @returns {Version | undefined}
 */
export function getLocalVersionForPath(
  folderPath: string,
): Version | undefined {
  return getLocalVersions().find((v) => v.localPath === folderPath);
}

/**
 * Retrieves local Electron versions, configured by the user.
 *
 * @returns {Array<Version>}
 */
export function getLocalVersions(): Array<Version> {
  const versions = getVersions(VersionKeys.local, () => []);

  return versions;
}

/**
 * Saves local versions to localStorage.
 *
 * @param {Array<Version | RunnableVersion>} versions
 */
export function saveLocalVersions(versions: Array<Version | RunnableVersion>) {
  const filteredVersions = versions.filter((v) => {
    if (isElectronVersion(v)) {
      return v.source === VersionSource.local;
    }

    return true;
  });

  return saveVersions(VersionKeys.local, filteredVersions);
}

/**
 * Retrieves our best guess regarding the latest Electron versions. Tries to
 * fetch them from localStorage, then from a static releases.json file.
 *
 * @returns {Array<Version>}
 */
function getReleasedVersions(): Array<Version> {
  return getVersions(VersionKeys.known, () => {
    return releasesJSON as Array<Version>;
  });
}

/**
 * Helper to check if this version is from a released major branch.
 *
 * This way when we have a local version of Electron like '999.0.0'
 * we'll know to not try & download 999-x-y.zip from GitHub :D
 *
 * @param {number} major - Electron major version number
 * @returns {boolean} true if there are releases with that major version
 */
export function isReleasedMajor(major: number) {
  const prefix = `${major}.`;
  return getReleasedVersions().some((ver) => ver.version.startsWith(prefix));
}

/**
 * Saves known versions to localStorage.
 *
 * @param {Array<Version>} versions
 */
function saveKnownVersions(versions: Array<Version>) {
  return saveVersions(VersionKeys.known, versions);
}

/**
 * Fetch a list of released versions from electronjs.org.
 *
 * @returns {Promise<Version[]>}
 */
export async function fetchVersions(): Promise<Version[]> {
  const url = 'https://releases.electronjs.org/releases.json';
  const response = await window.fetch(url, {
    cache: 'no-store',
  });
  const data = (await response.json()) as { version: string }[];

  const versions: Version[] = data
    // Don't support anything older than 0.30 (Aug 2015).
    // The oldest version known to releases.json.org is 0.20,
    // Pre-0.24.0 versions were technically 'atom-shell' and cannot
    // be downloaded with @electron/get.
    .filter((ver) => !ver.version.startsWith('0.2'))
    .map(({ version }) => ({ version }));

  console.log(`Fetched ${versions.length} new Electron versions`);
  if (versions.length > 0) saveKnownVersions(versions);
  return versions;
}

/**
 * Is the given array an array of versions?
 *
 * @param {Array<any>} input
 * @returns {boolean}
 */
function isExpectedFormat(input: Array<any>): boolean {
  return input.every((entry) => !!entry.version);
}

/**
 * Migrates old versions, if necessary
 *
 * @param {Array<any>} input
 * @returns {Array<Version>}
 */
function migrateVersions(input: Array<any>): Array<Version> {
  return input
    .filter((item) => !!item)
    .map((item) => {
      const { tag_name, name, url } = item;

      if (!tag_name || !name || !url) return null;

      return {
        version: tag_name,
        name,
        localPath: url,
      };
    })
    .filter((item) => !!item) as Array<Version>;
}

function isElectronVersion(
  input: Version | RunnableVersion,
): input is RunnableVersion {
  return (input as RunnableVersion).source !== undefined;
}

export function getOldestSupportedMajor(): number | undefined {
  const NUM_BRANCHES = parseInt(process.env.NUM_STABLE_BRANCHES || '') || 4;

  return getReleasedVersions()
    .filter((ver) => ver.version.endsWith('.0.0'))
    .map((ver) => Number.parseInt(ver.version))
    .sort((a, b) => a - b)
    .slice(-NUM_BRANCHES)
    .shift();
}
