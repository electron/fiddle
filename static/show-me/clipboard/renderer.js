window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
