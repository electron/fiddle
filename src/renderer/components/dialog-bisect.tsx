import { Button, Dialog, Label, MenuItem } from '@blueprintjs/core';
import { Select } from '@blueprintjs/select';
import { observer } from 'mobx-react';
import * as React from 'react';

import { ElectronVersion } from '../../interfaces';
import { Bisector } from '../bisect';
import { AppState } from '../state';
import { filterItem, getItemIcon, getVersionsFromAppState, renderItem } from './commands-version-chooser';

const ElectronVersionSelect = Select.ofType<ElectronVersion>();

export interface BisectDialogProps {
  appState: AppState;
}

export interface BisectDialogState {
  startIndex?: number;
  endIndex?: number;
  allVersions: Array<ElectronVersion>;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<BisectDialogProps, BisectDialogState>}
 */
@observer
export class BisectDialog extends React.Component<BisectDialogProps, BisectDialogState> {
  constructor(props: BisectDialogProps) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onBeginSelect = this.onBeginSelect.bind(this);
    this.onEndSelect = this.onEndSelect.bind(this);

    const allVersions = getVersionsFromAppState(this.props.appState);
    this.state = { allVersions };
  }

  public onBeginSelect(version: ElectronVersion) {
    this.setState({ startIndex: this.state.allVersions.indexOf(version) });
  }

  public onEndSelect(version: ElectronVersion) {
    this.setState({ endIndex: this.state.allVersions.indexOf(version) });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { endIndex, startIndex, allVersions } = this.state;
    const { appState } = this.props;

    if (endIndex === undefined || startIndex === undefined) {
      return;
    }

    const bisectRange = allVersions
      .slice(endIndex, startIndex + 1)
      .reverse();

    appState.Bisector = new Bisector(bisectRange);
    const initialBisectPivot = appState.Bisector.getCurrentVersion().version;
    appState.setVersion(initialBisectPivot);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isBisectDialogShowing = false;
  }

  get buttons() {
    const canSubmit =
      !!this.state.startIndex &&
      !!this.state.endIndex &&
      this.state.startIndex > this.state.endIndex;

    return [
      (
        <Button
          icon='play'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
          text='Begin'
        />
      ), (
        <Button
          icon='cross'
          key='cancel'
          onClick={this.onClose}
          text='Cancel'
        />
      )
    ];
  }

  public render() {
    const { isBisectDialogShowing } = this.props.appState;
    const { startIndex, endIndex, allVersions } = this.state;

    return (
      <Dialog
        isOpen={isBisectDialogShowing}
        onClose={this.onClose}
        title='Start a bisect session'
        className='dialog-add-version'
      >
        <div className='bp3-dialog-body'>
          <Label>
            Earliest Version
            <ElectronVersionSelect
              filterable={true}
              items={allVersions}
              itemRenderer={renderItem}
              itemPredicate={filterItem}
              onItemSelect={this.onBeginSelect}
              noResults={<MenuItem disabled={true} text='No results.' />}
            >
              <Button
                text={startIndex ? `v${allVersions[startIndex].version}` : ``}
                icon={startIndex ? getItemIcon(allVersions[startIndex]) : 'small-minus'}
                fill={true}
              />
            </ElectronVersionSelect>
          </Label>
          <Label>
            Latest Version
            <ElectronVersionSelect
              filterable={true}
              items={allVersions.slice(0, startIndex!)}
              itemRenderer={renderItem}
              itemPredicate={filterItem}
              onItemSelect={this.onEndSelect}
              noResults={<MenuItem disabled={true} text='No results.' />}
              disabled={!startIndex}
            >
              <Button
                text={endIndex ? `v${allVersions[endIndex].version}` : ``}
                icon={endIndex ? getItemIcon(allVersions[endIndex]) : 'small-minus'}
                fill={true}
                disabled={!startIndex}
              />
            </ElectronVersionSelect>
          </Label>
        </div>
        <div className='bp3-dialog-footer'>
          <div className='bp3-dialog-footer-actions'>
            {this.buttons}
          </div>
        </div>
      </Dialog>
    );
  }
}
