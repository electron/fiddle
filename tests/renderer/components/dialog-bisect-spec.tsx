import { shallow } from 'enzyme';
import * as React from 'react';
import { ElectronVersionSource, ElectronVersionState } from '../../../src/interfaces';
import { Bisector } from '../../../src/renderer/bisect';
import { BisectDialog } from '../../../src/renderer/components/dialog-bisect';
import { ElectronReleaseChannel } from '../../../src/renderer/versions';

jest.mock('../../../src/renderer/bisect');

describe('BisectDialog component', () => {
  let store: any;

  const generateVersionRange = (rangeLength: number) =>
    (new Array(rangeLength)).fill(0).map((_, i) => ({
      state: ElectronVersionState.ready,
      version: `${i + 1}.0.0`,
      source: ElectronVersionSource.local
    }));

  beforeEach(() => {
    store = {
      versions: generateVersionRange(5),
      versionsToShow: [ElectronReleaseChannel.stable],
      statesToShow: [ElectronVersionState.ready],
      setVersion: jest.fn()
    };
  });

  it('renders', () => {
    const wrapper = shallow(<BisectDialog appState={store} />);
    // start and end selected
    wrapper.setState({
      startIndex: 3,
      endIndex: 0,
      allVersions: generateVersionRange(5)
    });
    expect(wrapper).toMatchSnapshot();

    // no selection
    wrapper.setState({
      startIndex: undefined,
      endIndex: undefined,
      allVersions: generateVersionRange(5)
    });
    expect(wrapper).toMatchSnapshot();

    // only start selected
    wrapper.setState({
      startIndex: 3,
      endIndex: undefined,
      allVersions: generateVersionRange(5)
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('onBeginSelect()', () => {
    it('sets the begin version', () => {
      const wrapper = shallow(<BisectDialog appState={store} />);
      const instance: BisectDialog = wrapper.instance() as any;

      expect(instance.state.startIndex).toBeUndefined();
      instance.onBeginSelect(store.versions[2]);
      expect(instance.state.startIndex).toBe(2);
    });
  });

  describe('onEndSelect()', () => {
    it('sets the end version', () => {
      const wrapper = shallow(<BisectDialog appState={store} />);
      const instance: BisectDialog = wrapper.instance() as any;

      expect(instance.state.endIndex).toBeUndefined();
      instance.onEndSelect(store.versions[2]);
      expect(instance.state.endIndex).toBe(2);
    });
  });

  describe('onSubmit()', () => {
    it('initiates a bisect instance and sets a version', async () => {
      const version = '1.0.0';
      (Bisector as jest.Mock).mockImplementation(() => {
        return {
          getCurrentVersion: () => ({ version })
        };
      });

      const versions = generateVersionRange(5);

      const wrapper = shallow(<BisectDialog appState={store} />);
      wrapper.setState({
        startIndex: 4,
        endIndex: 0,
        allVersions: versions
      });

      const instance: BisectDialog = wrapper.instance() as any;
      await instance.onSubmit();
      expect(Bisector).toHaveBeenCalledWith(versions.slice(0, 5).reverse());
      expect(store.Bisector).toBeDefined();
      expect(store.setVersion).toHaveBeenCalledWith(version);
    });

    it('does nothing if endIndex or startIndex are falsy', async () => {
      const wrapper = shallow(<BisectDialog appState={store} />);

      wrapper.setState({
        startIndex: undefined,
        endIndex: 0
      });
      const instance1: BisectDialog = wrapper.instance() as any;
      await instance1.onSubmit();
      expect(Bisector).not.toHaveBeenCalled();

      wrapper.setState({
        startIndex: 4,
        endIndex: undefined
      });
      const instance2: BisectDialog = wrapper.instance() as any;
      await instance2.onSubmit();
      expect(Bisector).not.toHaveBeenCalled();

    });
  });
});
