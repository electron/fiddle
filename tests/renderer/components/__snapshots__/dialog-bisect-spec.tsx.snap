// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BisectDialog component renders 1`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  className="dialog-add-version"
  isOpen={false}
  onClose={[Function]}
  title="Start a bisect session"
>
  <div
    className="bp3-dialog-body"
  >
    <Component>
      Earliest Version
      <Blueprint3.Select
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "4.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "5.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          fill={true}
          icon="saved"
          text="v4.0.0"
        />
      </Blueprint3.Select>
    </Component>
    <Component>
      Latest Version
      <Blueprint3.Select
        disabled={false}
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          disabled={false}
          fill={true}
          icon="small-minus"
          text=""
        />
      </Blueprint3.Select>
    </Component>
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        disabled={true}
        icon="play"
        key="submit"
        onClick={[Function]}
        text="Begin"
      />
      <Blueprint3.Button
        icon="cross"
        key="cancel"
        onClick={[Function]}
        text="Cancel"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;

exports[`BisectDialog component renders 2`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  className="dialog-add-version"
  isOpen={false}
  onClose={[Function]}
  title="Start a bisect session"
>
  <div
    className="bp3-dialog-body"
  >
    <Component>
      Earliest Version
      <Blueprint3.Select
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "4.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "5.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          fill={true}
          icon="small-minus"
          text=""
        />
      </Blueprint3.Select>
    </Component>
    <Component>
      Latest Version
      <Blueprint3.Select
        disabled={true}
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "4.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "5.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          disabled={true}
          fill={true}
          icon="small-minus"
          text=""
        />
      </Blueprint3.Select>
    </Component>
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        disabled={true}
        icon="play"
        key="submit"
        onClick={[Function]}
        text="Begin"
      />
      <Blueprint3.Button
        icon="cross"
        key="cancel"
        onClick={[Function]}
        text="Cancel"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;

exports[`BisectDialog component renders 3`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  className="dialog-add-version"
  isOpen={false}
  onClose={[Function]}
  title="Start a bisect session"
>
  <div
    className="bp3-dialog-body"
  >
    <Component>
      Earliest Version
      <Blueprint3.Select
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "4.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "5.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          fill={true}
          icon="saved"
          text="v4.0.0"
        />
      </Blueprint3.Select>
    </Component>
    <Component>
      Latest Version
      <Blueprint3.Select
        disabled={false}
        filterable={true}
        itemPredicate={[Function]}
        itemRenderer={[Function]}
        items={
          Array [
            Object {
              "source": "local",
              "state": "ready",
              "version": "1.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "2.0.0",
            },
            Object {
              "source": "local",
              "state": "ready",
              "version": "3.0.0",
            },
          ]
        }
        noResults={
          <Blueprint3.MenuItem
            disabled={true}
            multiline={false}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="No results."
          />
        }
        onItemSelect={[Function]}
      >
        <Blueprint3.Button
          disabled={false}
          fill={true}
          icon="small-minus"
          text=""
        />
      </Blueprint3.Select>
    </Component>
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        disabled={true}
        icon="play"
        key="submit"
        onClick={[Function]}
        text="Begin"
      />
      <Blueprint3.Button
        icon="cross"
        key="cancel"
        onClick={[Function]}
        text="Cancel"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;
